docs(mirror-node-integration): add concept doc and connecting a Mirror Node to a Block Node document (#2837)
hiero-ledger/hiero-block-node: docs/block-node/operations/connecting-a-mirror-node-to-a-block-node.md

| High latency between block production and Mirror Node receipt                                | Live queue is polled at up to `MAX_LIVE_POLL_DELAY = 500 ms`.                                                              | This is the worst-case poll latency in the current implementation.                                                                                                                                                       |
| Session fails shortly after reconnect with `NOT_AVAILABLE (6)`                               | Mirror Node reconnected before the Block Node re-indexed the requested range after a restart.                              | Add a short delay and re-query `serverStatus` before each reconnect (already handled by the Mirror Node's readmit logic via `readmitDelay`).                                                                             |

> | Cutover does not switch to block-stream source                                               | `hiero.mirror.importer.block.cutover.enabled` is `false`, or the network has not reached `cutover.firstStage.hapiVersion`. | Set `cutover.enabled=true`; confirm the network HAPI version meets `cutover.firstStage.hapiVersion`.                                                                                                                     |

Reviewer: ```suggestion
| Cutover does not switch to block-stream source                                               | `hiero.mirror.importer.block.cutover.enabled` is `false`, or the network has not reached `cutover.hapiVersion`. | Set `cutover.enabled=true`; confirm the network HAPI version meets `cutover.hapiVersion`.                                                                                                                     |
```

For the rationale, subscription model, status codes, and the gap / reconnection behaviour Mirror Node operators must understand, see the companion [Mirror Node Integration](../mirror-node-integration.md) concept doc.

## Overview

A Mirror Node opens a long-lived gRPC server-streaming call to each Block Node's `BlockStreamSubscribeService.subscribeBlockStream` RPC. The Block Node streams `SubscribeStreamResponse` messages containing block items, terminates each block with an `end_of_block` marker, and closes the stream with a single terminal `status` message.

The Mirror Node ships with built-in multi-Block-Node support: configure one or more Block Nodes under `hiero.mirror.importer.block.nodes[]` and the Mirror Node selects between them by priority and measured latency, failing over automatically if a node becomes inactive.

This guide shows how to:

1. Confirm each target Block Node is reachable and serving the desired block range.
2. Configure the Mirror Node to subscribe — including the migration cutover toggle.
3. Verify the connection through logs, metrics, and the Block Node's `serverStatus` endpoint.
4. Diagnose the most common failure modes.

> > **Most Mirror Node operators consume the block stream from a public Block Node and do not run their own.** The exception is an operator running many Mirror Nodes, who may run a Tier 2 Block Node as a local buffer that pre-validates and redistributes the stream to multiple Mirror Nodes from a single connection upstream. This guide covers both cases; sections specific to running your own Block Node are marked accordingly.

Reviewer 2: Can we re-word this so it's not misinterpreted as a prescription of ways MN operator could/should run BNs.
If anything just say this covers MNs pointing to Tier 1 or Tier 2 BNs

Reviewer 2: Will push out so it's not blocking

| `hiero.mirror.importer.block.stream.readmitDelay`         | `1m`                    | How long an inactive Block Node stays out before being retried.                                                                                 |
| `hiero.mirror.importer.block.stream.responseTimeout`      | `400ms`                 | `serverStatus` request timeout.                                                                                                                 |

#### Migration cutover

During the records-to-block-streams cutover ([HIP-1193](https://github.com/hiero-ledger/hiero-improvement-proposals/blob/main/HIP/hip-1193.md)), the Mirror Node can switch automatically from record-file ingestion to block-stream subscription as the network rolls forward:

|                           Property                           | Default  |                                    Effect                                     |
|--------------------------------------------------------------|----------|-------------------------------------------------------------------------------|

> | `hiero.mirror.importer.block.cutover.enabled`                | `false`  | Enables the auto-switch. Keep `false` until the operator is ready to migrate. |

Reviewer: In release 0.155, we changed back the default to unset and this make it dependent on the network. The effective value when this property is unset is as follows

- `true` for `mainnet` and `testnet`
- `false` for all other networks

If the property is set, it overrides the network default.

```suggestion
| `hiero.mirror.importer.block.cutover.enabled`                |   | Enables the auto-switch. Keep `false` until the operator is ready to migrate. |
```

We also added a new hapiVersion config property

```
| `hiero.mirror.importer.block.cutover.hapiVersion`                               | 0.76.0                                               | The HAPI version following which the final cutover will happen.                                                                                                                                                                                                    |
```

So the default cutover detection will only kick in when the network runs 0.76 release.

| `hiero.mirror.importer.block.cutover.firstStage.enabled`     | `false`  | Enables first-stage cutover with latency safety checks.                       |
| `hiero.mirror.importer.block.cutover.firstStage.hapiVersion` | `0.75.0` | Minimum HAPI version that enables first-stage cutover.                        |

> | `hiero.mirror.importer.block.cutover.firstStage.maxLatency`  | `4s`     | If exceeded during the latency check, falls back to record stream.            |

Reviewer: I'd prefer not to recommend these to third-party operators.

```suggestion
```

| `hiero.mirror.importer.block.cutover.threshold`              | `16s`    | Time to wait before switching between block stream and record stream.         |

> For migration, the recommended sequence is: ensure record-file ingestion is healthy → set `block.enabled=true` and `sourceType=AUTO` to begin block-stream verification in parallel → flip `block.cutover.enabled=true` to authorize the switch once the network reaches the cutover release.

Reviewer: cutover detection and handling is automatic, we shouldn't recommend any config changes to operators.

```suggestion
```

The most important config is setting block node server host and optionally port. I'm not sure if we have finalized on the approach: HIP-1137 registered node so it's automatic, or manual configuration, or a later mirrornode release with a list of default block node servers.

- **Expected output** (active node with blocks ingested):

  ```json
  {
    "firstAvailableBlock": "1",
    "lastAvailableBlock": "123456",
    "onlyLatestState": false
  }
  ```
- **Expected output** (freshly started node, no blocks yet):

  ```json
  {
    "firstAvailableBlock": "18446744073709551615",
    "lastAvailableBlock": "18446744073709551615",
    "onlyLatestState": false
  }
  ```

  When both values are `uint64_max`, the Block Node has not yet ingested any blocks. The Mirror Node will receive `NOT_AVAILABLE (6)` if it tries to subscribe now.

  Without `-emit-defaults`, `grpcurl` elides `"onlyLatestState": false` from the output; both forms are semantically equivalent.

### Step 2: Configure the Mirror Node and restart

1. Edit the Mirror Node importer's `application.yml` (or equivalent override).
2. Set the [required properties](#required) and add a `nodes[]` entry per Block Node, as shown in the example above.

> 3. If you are mid-migration and want the Mirror Node to switch from record-file ingestion to block-stream subscription automatically, also set [`hiero.mirror.importer.block.cutover.enabled=true`](#migration-cutover).

Reviewer: This isn't needed.

```suggestion
```

The Mirror Node ships with built-in multi-Block-Node support: configure one or more Block Nodes under `hiero.mirror.importer.block.nodes[]` and the Mirror Node selects between them by priority and measured latency, failing over automatically if a node becomes inactive.

This guide shows how to:

1. Confirm each target Block Node is reachable and serving the desired block range.
2. Configure the Mirror Node to subscribe — including the migration cutover toggle.
3. Verify the connection through logs, metrics, and the Block Node's `serverStatus` endpoint.
4. Diagnose the most common failure modes.

## Prerequisites

Before you begin, ensure you have:

- A running Block Node deployment using one of the following methods:
  - [**Manual Single-Node Kubernetes Deployment**](./single-node-k8s-deployment.md)
  - [**Solo Provisioner Single-Node Kubernetes Deployment**](./solo-weaver-single-node-k8s-deployment.md)

>   - [**Block Node Dev Quickstart (Docker)**](../quickstart.md) — for local development and testing.

Reviewer 3: Here we should add a reference to public block nodes

```suggestion
- A running Block Node deployment using one of the following methods:
  - [**Manual Single-Node Kubernetes Deployment**](./single-node-k8s-deployment.md)
  - [**Solo Provisioner Single-Node Kubernetes Deployment**](./solo-weaver-single-node-k8s-deployment.md)
  - [**Block Node Dev Quickstart (Docker)**](../quickstart.md) — for local development and testing.
  - [**Block Nodes Published on the Network**](https://hips.hedera.com/all#hip-1137) - For production use with production networks.
```

We might also want to include a paragraph explaining up front that _in production_ mirror node operators are not generally expected to run their own block node. The exception being that an operator with _many_ mirror nodes should run a "tier 2" block node to ease load on tier 1 and also provide local buffer and pre-validation.

- A running Mirror Node deployment ready to be reconfigured to consume from a Block Node:
  - [**Mirror Node Installation Guide**](https://github.com/hiero-ledger/hiero-mirror-node/blob/main/docs/installation.md) — local or Docker Compose install paths.
  - [**Mirror Node Configuration Reference**](https://github.com/hiero-ledger/hiero-mirror-node/blob/main/docs/configuration.md) — full property reference for the Mirror Node services.
- Network connectivity between the Mirror Node host and each Block Node host on the gRPC port (default `40840`).
- A gRPC client capable of HTTP/2 server-streaming calls:
  - The Mirror Node's built-in gRPC stack (for production integration), **or**

>   - [**`grpcurl`**](https://github.com/fullstorydev/grpcurl) — for ad-hoc verification from a shell (optional but recommended). Install via `brew install grpcurl` on macOS or your distribution's package manager on Linux.

Reviewer 3: Perhaps also note Windows install:

```suggestion
  - [**`grpcurl`**](https://github.com/fullstorydev/grpcurl) — for ad-hoc verification from a shell (optional but recommended). Install via `brew install grpcurl` on macOS, following the [official guide](https://grpcurl.com/how-do-i-install-grpcurl-on-my-system/) on Windows, or your distribution's package manager on Linux.
```

|------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Block Node version     | A release matching the proto and configuration referenced in this guide. [NEEDS-VERIFICATION — confirm the minimum supported release once a tagged version after `0.33.0` is available.] |
| gRPC port reachable    | Default `40840`; configurable via `server.port`.                                                                                                                                         |
| Metrics port reachable | Default `16007` for verification (optional but recommended).                                                                                                                             |
| Block availability     | The Block Node must have ingested at least `start_block_number` before the Mirror Node subscribes. Query `serverStatus` to confirm.                                                      |

> | Transport              | HTTP/2 over plaintext, or TLS terminated at infrastructure (load balancer, ingress). The Block Node process itself does not terminate TLS.                                               |

Reviewer 3: Given that we don't expect mirror node operators to run their own block node in most cases, perhaps it would be better to refer them to the block node documentation here?

| Proto definitions | `block_stream_subscribe_service.proto`, `node_service.proto`, and `shared_message_types.proto` from [`protobuf-sources/src/main/proto/block-node/api/`](https://github.com/hiero-ledger/hiero-block-node/tree/main/protobuf-sources/src/main/proto/block-node/api). For ad-hoc `grpcurl` use, the matching versioned bundle from the [Block Node releases](https://github.com/hiero-ledger/hiero-block-node/releases) page is the easiest source — see [Step 1](#step-1-confirm-each-block-node-is-reachable-and-serving-blocks). |
| gRPC reflection   | The Block Node does **not** enable gRPC server reflection on the public port. Clients must supply protobuf descriptors explicitly.                                                                                                                                                                                                                                                                                                                                                                                                |

## Configuration

> ### Block Node settings that affect subscribers

Reviewer 3: As above, it might be better to refer to the block node documentation here for brevity.

  -import-path . \
  -proto block-node/api/block_stream_subscribe_service.proto \
  -d '{"start_block_number": "1", "end_block_number": "18446744073709551615"}' \
  <BLOCK_NODE_HOST>:40840 \
  org.hiero.block.api.BlockStreamSubscribeService/subscribeBlockStream
```

- **Expected behaviour on a Block Node that has ingested blocks**: `grpcurl` prints a continuous stream of `SubscribeStreamResponse` messages alternating between `block_items` (batched block data) and `end_of_block` (one per completed block). The stream remains open until you cancel with `Ctrl-C` or the Block Node returns a terminal `status`.
- **Expected behaviour on a freshly started Block Node with no blocks**: a single terminal status, then the stream closes:

  ```json
  { "status": "NOT_AVAILABLE" }
  ```

### Step 4: Handle disconnects and gaps

The Block Node closes the stream when the finite range is fully served, when an internal error occurs, or when the client disconnects. The Mirror Node handles reconnection automatically: it will retry against the highest-priority reachable Block Node, governed by `maxSubscribeAttempts` and `readmitDelay`.

Two operator-visible patterns are worth knowing:

> - **Gap in `end_of_block.block_number`**: the Mirror Node fell behind, the Block Node trimmed whole blocks from the live-queue head, and the Mirror Node must reconnect to backfill the missing range from history. The Mirror Node detects this and re-subscribes with `start_block_number = (last_committed_block + 1)`. If the gap is not available on the current Block Node (terminal `NOT_AVAILABLE`), failover to a higher-priority Tier-1 archive node is required — configure such a node as a low-priority entry under `nodes[]` so the Mirror Node can fall over automatically.

Reviewer 3: This is slightly inaccurate.

```suggestion
- **Gap in `end_of_block.block_number`**: The flow of data from consensus node to block node arrived out-of-order or required a resend due to verification or persistence failure. The Mirror Node may reconnect to backfill the missing range from history. The Mirror Node detects this and re-subscribes with `start_block_number = (last_committed_block + 1)`. If the gap is not available on the current Block Node (terminal `NOT_AVAILABLE`), failover to a higher-priority Tier-1 archive node is required — configure such a node as a low-priority entry under `nodes[]` so the Mirror Node can fall over automatically.
```

- **Repeated `ERROR (3)` from one Block Node**: the Block Node is failing internally. The Mirror Node will mark it inactive after `maxSubscribeAttempts` and try the next configured node.

> See [Slow consumers see gaps, not closed streams](../mirror-node-integration.md#slow-consumers-see-gaps-not-closed-streams) in the concept doc for the rationale.

Reviewer 3: The "slow consumers" part is misleading here.  Slow consumers see data fed from history, not gaps. The subscribers see gaps because the stream is _unverified_.

## Verification

### Verify on each Block Node side

#### Logs

When a session ends in error, the Block Node logs at `INFO` level (the `%(,d` format specifier expands to the numeric client identifier):

```
Subscriber session <clientId> failed due to <cause>.
```

When a session ends with `SUCCESS`, it logs at `TRACE` level:

```
Subscriber session <clientId> completed successfully.
```

Enable `TRACE` for the `org.hiero.block.node.stream.subscriber` logger if you want positive confirmation per session.

#### Metrics

The Block Node exposes Prometheus-format metrics on the default metrics endpoint:

```bash
curl -s http://<BLOCK_NODE_HOST>:16007/metrics | grep blocknode_subscriber
```

- **Expected output** with one Mirror Node connected:

  ```
  # TYPE blocknode_subscriber_open_connections gauge
  blocknode_subscriber_open_connections 1
  # TYPE blocknode_subscriber_errors counter
  blocknode_subscriber_errors_total 0
  ```

>   The counter is exposed with the Prometheus-conventional `_total` suffix even though the underlying metric is registered as `subscriber_errors`. The `_open_connections` gauge is updated lazily: a closed session's decrement is processed only when the next subscriber attempts to connect. In low-traffic windows the gauge can appear stuck on the previous value. Treat the gauge as approximate; use the Mirror Node side (last committed block, reconnect rate) and infrastructure-level connection counts (load balancer, ingress) for precise observation.

Reviewer 3: This is specific to entities running a block node; which is not necessarily expected of mirror node operators.
Perhaps move this to an appendix or other more clearly optional section?

| Repeated terminal `ERROR (3)` from a single Block Node                                       | Block Node internal failure.                                                                                               | Check that Block Node's logs at `INFO` for `failed due to ...`; review its health (CPU, memory, disk). The Mirror Node will mark this node inactive after `maxSubscribeAttempts`.                                                         |
| All configured Block Nodes marked inactive                                                   | Network reachability problem, or all Block Nodes simultaneously unhealthy.                                                 | Verify host/port for each `nodes[]` entry; check that `requiresTls` matches the actual termination setup; inspect each Block Node's metrics endpoint.                                                                                     |

> | `blocknode_subscriber_open_connections` does not increment after Mirror Node connects        | Connection is not reaching the Block Node.                                                                                 | Re-check firewall rules; verify the Mirror Node is hitting the correct host and port.                                                                                                                                                     |

Reviewer 3: Slight wording improvement:

```suggestion
| `blocknode_subscriber_open_connections` does not increment after Mirror Node connects        | Connection is not reaching the Block Node.                                                                                 | Re-check firewall rules; verify the Mirror Node is connecting to the correct host and port.                                                                                                                                                     |
```

> | **Gap in `end_of_block.block_number` after going live**                                      | The Mirror Node fell behind and the Block Node trimmed whole blocks from the live queue head.                              | The Mirror Node will reconnect to backfill from history. To reduce the rate of trims, increase `subscriber.liveQueueSize` and/or `subscriber.minimumLiveQueueCapacity` on the Block Node, or improve the Mirror Node's commit throughput. |

Reviewer 3: The cause here is not a slow mirror node, and changing these values will have no impact on gaps in block number.

```suggestion
| **Gap in `end_of_block.block_number` after going live**                                      | The live block stream experienced interruptions or errors and blocks were received out of order or were resent.            | The Mirror Node will reconnect to backfill from history.                                                                                                                                                                                  |
```
